### PR TITLE
using args-mutation annotation to support inplace op in torch-blade pipline

### LINKED ISSUE #1218: support inplace operator in TorchBlade
Pytorch inplace operators are widely used in model training/inference, just as KV cache in LLM, TorchBlade should support these
inplace operator.
 
Just as the following simple PyTorch program with in-place add operator `aten.add_`:
``` python
def func(add : Tensor, value : Tensor) -> Tensor:
  add.add_(value)
  return add
```
torchdynamo can trace the above program into value-semantic operators and insert `aten._copy` at the back:

``` python
add: f32[8, 64] = torch.ops.aten.add.Tensor(arg0_1, arg1_1);  arg1_1 = None
# No stacktrace found for the following nodes
copy_: f32[8, 64] = torch.ops.aten.copy_.default(arg0_1, add);  arg0_1 = None
```

The naive operation in buffer level can be represent as:
``` mlir
lmhlo.add %arg0, %arg1, %1
lmhlo.copy %1, %arg0
```

but in actuality the `lmhlo.copy ` causes overhead because of the necessary d2d memcpy, we can remove this copy and store the result element on the input buffer:

``` mlir
lmhlo.add %arg0, %arg1, %arg0
```

### PR DESCRIPTION
**Pull request by Yancey1989**
fixed #1218

### CODE REVIEW

File: pytorch_blade/pytorch_blade/torch-mlir/lib/Dialect/TorchConversion/Transforms/ApplyValueSemantics.cpp [hunk 1]
```diff
@@ -89,6 +90,16 @@ class ApplyValueSemanticsPass
       op->erase();
     }
 
+    // fix NonValueTensor type operator
+    func.walk([&](Operation* op) {
+      if (isa<OverwriteTensorContentsOp>(op)) {
+        rewriter.setInsertionPoint(op);
+        auto nonValueTensor = rewriter.create<CopyToNonValueTensorOp>(
```
Comment (qiuxiafei) [inline pytorch_blade/pytorch_blade/torch-mlir/lib/Dialect/TorchConversion/Transforms/ApplyValueSemantics.cpp:97]: Why do we need to convert `OverwriteTensorContentsOp` to `CopyToNonValueTensorOp` and then to `disc.ArgsMutation`? Is it possible to skip `CopyToNonValueTensorOp`?

**Review by qiuxiafei — APPROVED** [review]
LGTM with minor comments.

File: pytorch_blade/pytorch_blade/torch-mlir/lib/Dialect/TorchConversion/Transforms/ApplyValueSemantics.cpp (hunk 1 above)
Comment (Yancey1989) [reply pytorch_blade/pytorch_blade/torch-mlir/lib/Dialect/TorchConversion/Transforms/ApplyValueSemantics.cpp:97]: Maybe `CopyToNonValueTensorOp ` is required, the function `reduceTensorConversions(func)` at L79 removs all Tensor conversations and convert all `!tensor` to `!vtensor`,  but `OverwriteTensorContentsOp ` operand requires `!tensor` , so we need to insert `CopyToNonValueTensorOp` to make `OverwriteTensorContentsOp ` validation after `reduceTensorConversions`.

File: pytorch_blade/pytorch_blade/torch-mlir/lib/Conversion/TorchToMhlo/DiscTorchToMhlo.cpp [hunk 2]
```diff
@@ -765,7 +765,7 @@ LogicalResult ConvertAtenOp<AtenSizeIntOp>::matchAndRewrite(
     OpAdaptor adaptor,
     ConversionPatternRewriter& rewriter) const {
   // Not a tensor type.
-  auto selfType = adaptor.getSelf().getType().dyn_cast<TensorType>();
+  auto selfType = adaptor.getSelf().getType().cast<RankedTensorType>();
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply pytorch_blade/pytorch_blade/torch-mlir/lib/Conversion/TorchToMhlo/DiscTorchToMhlo.cpp:768]: you're right, will polish this in the next PR.

**Pull request merged**